Fix the mismatching default configs in integration tests

jihoonson opened the pull request:
Fixes https://github.com/NVIDIA/spark-rapids/issues/11134.

Default values of some configs used in the integration test by default has gone stale as they haven't updated when their defaults changed. This is because we have a separated list of default configs used by the integration test, which should be manually updated along with code change. This PR fixes this problem by dumping all default configs in a file and letting the integration test use it. A new interface, `getDefault`, has been added in `ConfEntry`. A new `DumpDefaultConfigs` class uses this API to dump all default configs in a file. In the integration test, those default configs are loaded when the spark session is initialized.

gerashegalov commented on integration_tests/pom.xml:
```diff
@@ -180,6 +180,36 @@
                 <artifactId>exec-maven-plugin</artifactId>
                 <executions>
+                    <execution>
+                        <id>clean previously populated config file</id>
```
Although we have contradicting examples. Through out the diff: It will be easier to use if the `id` does not have spaces even if Maven allows it. Consider replacing spaces with `-`

jihoonson replied:
I did follow those contradicting examples even though I thought it's strange 🙂 We should fix those IDs that have spaces in them. I fixed the ones I added and another nearby.

gerashegalov commented on integration_tests/pom.xml:
```diff
@@ -180,6 +180,36 @@
+                        <configuration>
+                            <includeProjectDependencies>true</includeProjectDependencies>
+                            <mainClass>com.nvidia.spark.rapids.tests.DumpDefaultConfigs</mainClass>
+                            <classpathScope>provided</classpathScope>
```
Will it not work with the default [classpathScope](https://www.mojohaus.org/exec-maven-plugin/java-mojo.html#classpathScope) ?

jihoonson replied:
No, it doesn't. I guess it's because the `rapids-4-spark_*` dependency is `provided`.

gerashegalov commented on integration_tests/run_pyspark_from_build.sh:
```diff
@@ -15,6 +15,7 @@
 set -ex
 
 SCRIPTPATH="$( cd "$(dirname "$0")" >/dev/null 2>&1 ; pwd -P )"
+export SPARK_RAPIDS_DEFAULT_CONFIGS_PATH=${SCRIPTPATH}/target/spark-rapids-default-configs.json
```
I think we can just have `spark-session.py` construct this path without this env variable

jihoonson replied:
I did this because I was not sure how to get the right path in `spark-session.py` even when the python script is executed in a different directory than the spark-rapids root. Or in some very random directory. Any idea how to do it in that python file?

gerashegalov replied:
We can follow the pattern with other paths, add a CLI option https://github.com/NVIDIA/spark-rapids/blob/bdc563ee3b380ed25fe6a817a6a04085ea0005c9/integration_tests/conftest.py#L15

jihoonson replied:
Hmm with that approach, should the user provide the correct path? I'm not sure how it is better than the current approach. The current one automatically locates the JSON file regardless of the CWD.

jihoonson replied:
Oh did you mean that we could pass the right path in the pom file if the pytest accepts the JSON file path as its argument? Yeah that would work.

gerashegalov replied:
the default will be ${SCRIPTPATH}/target/spark-rapids-default-configs.json 

but if the user needs to override it they can use a CLI param.

jihoonson replied:
Yeah, your suggestion makes sense now. I added a new option as you suggested.

From the scoping perspective, I appreciate your input, but this seems a bit out of the scope of the original PR. Your suggestion is to allow more flexible default configuration for integration test, which I agree could be useful, but it is not directly relevant to the original mismatching default configuration issue. I'd like to keep the scope as minimum as possible. We can always add such a feature later when we actually use it.

gerashegalov commented on integration_tests/src/main/python/spark_session.py:
```diff
@@ -67,7 +54,10 @@ def _set_all_confs(conf):
     for key, value in newconf.items():
         if _spark.conf.get(key, None) != value:
-            _spark.conf.set(key, value)
+            if isinstance(value, str):
```
Maybe only use strings in RapidsConf export to JSON to avoid casts on read?

jihoonson replied:
I think this is a bug in some JNI interface in `pysparkSparkConf`, not the JSON problem. It makes sense to me to wrap values here when we call the problematic API.

gerashegalov commented on integration_tests/src/main/scala/com/nvidia/spark/rapids/tests/DumpDefaultConfigs.scala:
```diff
@@ -0,0 +1,83 @@
+          case Format.PLAIN =>
+            val dos: DataOutputStream = new DataOutputStream(bos)
+            withResource(dos) { _ =>
+              allConfs.foreach({ case (k, v) =>
```
```suggestion
              allConfs.foreach { case (k, v) =>
```

gerashegalov commented on integration_tests/src/main/scala/com/nvidia/spark/rapids/tests/DumpDefaultConfigs.scala:
```diff
@@ -0,0 +1,83 @@
+            withResource(dos) { _ =>
+              allConfs.foreach({ case (k, v) =>
+                val valStr = v match {
+                  case some: Some[_] => some.getOrElse("")
```
```suggestion
                  case _: Some(optVal) => optVal.toString
```

jihoonson replied:
This matching doesn't seem to work? It seems to fail to infer the type of `optVal` (or maybe `Some`). The suggestion seems trivial to me, so I didn't try hard.

gerashegalov replied:
I did not test it maybe should have been some variation of

```
case Some(optVal) => 
```

jihoonson replied:
Thanks, updated as suggested.

gerashegalov commented on integration_tests/src/main/scala/com/nvidia/spark/rapids/tests/DumpDefaultConfigs.scala:
```diff
@@ -0,0 +1,83 @@
+              })
+            }
+          case Format.JSON =>
+            import org.json4s.jackson.Serialization.writePretty
+            import org.json4s.DefaultFormats
+            import java.nio.charset.StandardCharsets
```
nit: moving it to global imports will make it more readable I think

jihoonson replied:
Moved.

gerashegalov commented on integration_tests/src/main/scala/com/nvidia/spark/rapids/tests/DumpDefaultConfigs.scala:
```diff
@@ -0,0 +1,83 @@
+    val clazz = ShimReflectionUtils.loadClass("com.nvidia.spark.rapids.RapidsConf")
+    val m = clazz.getDeclaredMethod("getAllConfigsWithDefault")
+    val allConfs: Map[String, Any] = m.invoke(null).asInstanceOf[Map[String, Any]]
+    val fos: FileOutputStream = new FileOutputStream(outputPath)
+    withResource(fos) { _ =>
+      val bos: BufferedOutputStream = new BufferedOutputStream(fos)
+      withResource(bos) { _ =>
```
nit:
```suggestion
    withResource(new FileOutputStream(outputPath)) { fos =>
      withResource(new BufferedOutputStream(fos)) { bos =>
```

jihoonson replied:
Ah, this is how I should use this API. Thanks!

gerashegalov commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:
```diff
@@ -124,6 +124,7 @@ abstract class ConfEntry[T](val key: String, val converter: String => T, val doc
 
   def get(conf: Map[String, String]): T
   def get(conf: SQLConf): T
+  def getDefault(): T
```
It would rather make sense to omit the conf withtout defaults in the dump IMO

jihoonson replied:
So, I think this API should just return whatever the default is. Any business logic should not be involved at this level.

For `DumpDefaultConfigs`, it dumps even those configs without defaults because we would want to unset them for the integration test if they were set somehow.

jihoonson commented:
```@@ -1439,7 +1439,7 @@ This will force full Scala code rebuild in downstream modules.
                 <plugin>
                     <groupId>org.codehaus.mojo</groupId>
                     <artifactId>exec-maven-plugin</artifactId>
-                    <version>3.0.0</version>
+                    <version>3.3.0</version>
                 </plugin>
                 <plugin>
                     <groupId>org.apache.maven.plugins</groupId>
```

Uh, the CI fails with this error. What should I do about it? I bumped up this dependency version because the `provided` `classpathScope` has been introduced in that version.

jihoonson commented:
The CI error seems because of the missing version update in the 2.13 pom file for the corresponding dependency. I will fix it.

jlowe commented on integration_tests/run_pyspark_from_build.sh:
```diff
@@ -232,6 +232,7 @@ else
           "$TEST_ARGS"
           $RUN_TEST_PARAMS
           --junitxml=TEST-pytest-`date +%s%N`.xml
+          --default_configs_path="${SCRIPTPATH}/target/spark-rapids-default-configs.json"
```
```suggestion
          --default_configs_path="${TARGET_DIR}/spark-rapids-default-configs.json"
```

jlowe reviewed (commented):
Since this is adding a new dependency to the integration tests, it needs to update integration_tests/src/assembly/bin.xml to add the file to the integration tests tarball so CI test pipelines that do _not_ build from source can run the integration tests properly.

jihoonson commented:
@jlowe thanks for catching it. I added the new file in 8dbd354, but I'm not sure if I did it right. I opened up the `integration_tests/target/rapids-4-spark-integration-tests_2.12-24.08.0-SNAPSHOT-sparkxxx-jar-with-dependencies.jar`, but I don't see it there. Should the file be added somewhere else than this jar file?

jlowe commented on integration_tests/src/assembly/bin.xml:
```diff
@@ -50,6 +50,10 @@
             <outputDirectory>integration_tests</outputDirectory>
         </file>
+        <file>
+            <source>${spark.rapids.source.basedir}/integration_tests/target/spark-rapids-default-configs.json</source>
```
```suggestion
            <source>${project.build.outputDirectory}/spark-rapids-default-configs.json</source>
```

jihoonson replied:
Hmm I thought `project.build.outputDirectory` is `integration_tests/target/classes` not `integration_tests/target`?

gerashegalov replied:
you can use `${project.build.directory}`

jlowe replied:
Yes, sorry, picked the wrong build directory variable.  I just know that hardcoding `/target` is not what we should be doing, as there's a variable for that _somewhere_. :smile:

jihoonson replied:
Updated as suggested.

jihoonson commented:
```
java.lang.reflect.InaccessibleObjectException: Unable to make protected void java.net.URLClassLoader.addURL(java.net.URL) accessible: module java.base does not "opens java.net" to unnamed module @6808428c
    at java.lang.reflect.AccessibleObject.checkCanSetAccessible (AccessibleObject.java:354)
    at java.lang.reflect.AccessibleObject.checkCanSetAccessible (AccessibleObject.java:297)
    at java.lang.reflect.Method.checkCanSetAccessible (Method.java:200)
    at java.lang.reflect.Method.setAccessible (Method.java:194)
    at org.apache.commons.lang3.reflect.MethodUtils.invokeMethod (MethodUtils.java:839)
    at org.apache.commons.lang3.reflect.MethodUtils.invokeMethod (MethodUtils.java:802)
    at com.nvidia.spark.rapids.ShimLoader$.$anonfun$updateSparkClassLoader$2 (ShimLoader.scala:165)
    at com.nvidia.spark.rapids.ShimLoader$.$anonfun$updateSparkClassLoader$2$adapted (ShimLoader.scala:161)
    at scala.collection.immutable.List.foreach (List.scala:334)
    at com.nvidia.spark.rapids.ShimLoader$.$anonfun$updateSparkClassLoader$1 (ShimLoader.scala:161)
    at com.nvidia.spark.rapids.ShimLoader$.$anonfun$updateSparkClassLoader$1$adapted (ShimLoader.scala:160)
    at scala.Option.foreach (Option.scala:437)
    at com.nvidia.spark.rapids.ShimLoader$.updateSparkClassLoader (ShimLoader.scala:160)
    at com.nvidia.spark.rapids.ShimLoader$.getShimClassLoader (ShimLoader.scala:186)
    at com.nvidia.spark.rapids.ShimReflectionUtils$.loadClass (ShimReflectionUtils.scala:28)
    at com.nvidia.spark.rapids.tests.DumpDefaultConfigs$.main (DumpDefaultConfigs.scala:50)
```

Hmm so the CI is now failing with the error above. I think it's because of some missing JVM options, such as `--add-opens=java.base/java.net=ALL-UNNAMED` for java 17. I'm working on a fix.

gerashegalov commented on integration_tests/pom.xml:
```diff
@@ -78,6 +78,42 @@
+                <groupId>org.codehaus.mojo</groupId>
+                <artifactId>exec-maven-plugin</artifactId>
+                <executions>
+                    <execution>
+                        <id>clean-previously-populated-config-file</id>
+                        <phase>package</phase>
+                        <goals>
+                            <goal>exec</goal>
+                        </goals>
+                        <configuration>
+                            <executable>rm</executable>
+                            <arguments>
+                                <argument>-f</argument>
+                                <argument>${spark.rapids.source.basedir}/${rapids.module}/target/spark-rapids-default-configs.json</argument>
+                            </arguments>
+                        </configuration>
+                    </execution>
```
since we write under target. we should defer to the `clean` phase to remove the file

```suggestion
```

jihoonson replied:
👍 Makes sense. Removed this.

gerashegalov commented on integration_tests/run_pyspark_from_build.sh:
```diff
@@ -76,7 +76,7 @@ else
         LOWEST_PARQUET_JAR=$(echo -e "$MIN_PARQUET_JAR\n$PARQUET_HADOOP_TESTS" | sort -V | head -1)
         export INCLUDE_PARQUET_HADOOP_TEST_JAR=$([[ "$LOWEST_PARQUET_JAR" == "$MIN_PARQUET_JAR" ]] && echo true || echo false)
-        PLUGIN_JARS=$(echo "$TARGET_DIR"/../../dist/target/rapids-4-spark_*.jar)
+        PLUGIN_JARS=$(echo "$TARGET_DIR"/../../dist/rapids-4-spark_*.jar)
```
unrelated, If this is a real issue let us fix in a separate PR

jihoonson replied:
Hmm, thanks for catching this! It must be a mistake. I didn't intend to modify this line. Reverted it.

gerashegalov commented on integration_tests/pom.xml:
```diff
@@ -78,6 +78,42 @@
+                <executions>
+                    <execution>
+                        <id>clean-previously-populated-config-file</id>
+                        <phase>package</phase>
```
We need to make sure that this runs before the package phase such as generate-resources otherwise. It is hard to control order of executions across plugins in the same phase.

jihoonson replied:
I have removed this as you suggested.

gerashegalov replied:
I meant the actual json generation, probably misplaced my comment here

jihoonson commented:
As for the [CI failure with java 17](https://github.com/NVIDIA/spark-rapids/pull/11283#issuecomment-2264132655), to allow access to protected methods via reflection, I have two ideas:
1. We can add JVM options in `MAVEN_OPTS` to allow it. This `MAVEN_OPTS` can be set in the `mvn-verify-check.yml` so that the github action picks it up. We should update our doc as well to explain that those JVM options should be set to make a build with Java 17.
2. Instead of letting the maven process run `DumpDefaultConfigs`, we could spawn a new process by setting the goal of `exec-maven-plugin` to `exec` that executes a scala program. This will require making a jar with all the dependencies `DumpDefaultConfigs` needs, which seems only `scopt` currently besides the provided `rapids-4-spark_2.12` dependency.

I'm leaning towards to the option 1 only because it is easier. But the option 2 doesn't seem that hard either. Any opinions here?

jlowe commented:
> Any opinions here?

We already have an example of running a class as a program with all of its dependencies to generate files during the build which isn't doing any reflection shenanigans and IIUC is working with the JDK17 build.  See the [update_config_docs rule in dist/pom.xml](https://github.com/NVIDIA/spark-rapids/blob/branch-24.08/dist/pom.xml#L391-L424).  Any reason we can't do the same here?

gerashegalov commented:
+1 for following the pattern from update_config_docs. exec plugin executions are run in the Maven JVM, and requiring particular  MAVEN_OPTS is unnecessarily user-unfriendly. In Ant java task https://ant.apache.org/manual/Tasks/java.html we can set fork=true and pass the same args we already use for unit  tests ${extraJavaTestArgs} as `jvmargs`

jlowe commented on integration_tests/src/main/scala/com/nvidia/spark/rapids/tests/DumpDefaultConfigs.scala:
```diff
@@ -0,0 +1,81 @@
+    println(s"Dumping all spark-rapids configs and their defaults at ${outputPath}")
+
+    // We use the reflection as RapidsConf should be accessed via the shim layer.
+    val clazz = ShimReflectionUtils.loadClass("com.nvidia.spark.rapids.RapidsConf")
```
The logic for this class could be moved to RapidsConf which already generates files related to its settings.  This would be another file that is emitted when it is run, and integration_tests could pick up the file from this portion of the build.

jihoonson replied:
Hmm good suggestion. I think I added this class before imagining that this class would be useful only for integration test. Given that this class has been generalized enough (it takes the output format and the path as its params), I agree that `RapidsConf` is a better place to put this logic. I will make this change soon.

jihoonson commented:
I have spent some time to try out 3 approaches in the below and I think we should move forward with the Option 3 for now.

1. I tried to make a fat, executable jar with all dependencies for the `integration_tests` module. This jar could have been executed in the `package` phase, which would generate the default configs file. However, I was not able to make this work successfully as [the shim layer seems to require excluding all files for the jar packaging but attaching them later in the shim phase](https://github.com/NVIDIA/spark-rapids/blob/branch-24.10/integration_tests/pom.xml#L33-L34). I'm not sure though why this is required for the integration test.
2. I put the logic in `RapidsConf` and tried to invoke it in the `integration_tests` module. This doesn't seem to work because the `RapidsConf` class is shimmed and the default class loader cannot find it.
3. Another approach is what @jlowe suggested in [his comment](https://github.com/NVIDIA/spark-rapids/pull/11283#discussion_r1702204602). I moved the new logic to `RapidsConf` and generated the default configs file in the `package` phase for the `dist` module. The integration test picks up this file in the `dist/target` directory. This approach introduces a new, unnecessary dependency between the `dist` and the `integration_tests` modules.

I personally think that the Option 1 and the Option 2 are "better" than the Option 3 since the integration test generates a file it needs by itself. It does not introduce a new dependency between modules unnecessarily. The Option 2 is better than the Option 1 as it is simpler. However, I was not able to make either of them work. 

I think the unnecessary complexity in the current solution is being caused by https://github.com/NVIDIA/spark-rapids/issues/6556. I would rather move forward with the current ugly solution, which will promote the priority of https://github.com/NVIDIA/spark-rapids/issues/6556.

jlowe commented:
> This approach introduces a new, unnecessary dependency between the dist and the integration_tests modules.

The integration_tests module is supposed to be testing the output of the dist module, so I'm surprised there isn't already a dependency there.  What's the issue with there being a dependency between these modules?

jihoonson commented:
> > This approach introduces a new, unnecessary dependency between the dist and the integration_tests modules.
> 
> The integration_tests module is supposed to be testing the output of the dist module, so I'm surprised there isn't already a dependency there. What's the issue with there being a dependency between these modules?

The maven dependency between the two modules already exists. The dependency I referred to is a process dependency that the file used by the `integration_tests` module is generated by the `dist` module. I think this is not the best what we can do as the dependency is not obvious. You should go through the pom files in multiple modules to understand who creates the file how and who uses it.

jihoonson commented:
Finally it has passed all jobs in the CI 🙂 @jlowe @gerashegalov could you have another look?

jlowe commented on integration_tests/run_pyspark_from_build.sh:
```diff
@@ -223,6 +223,7 @@ else
 
     REPORT_CHARS=${REPORT_CHARS:="fE"} # default as (f)ailed, (E)rror
     STD_INPUT_PATH="$INPUT_PATH"/src/test/resources
+    DEFAULT_CONFIGS_PATH="${SCRIPTPATH}"/../dist/target/spark-rapids-default-configs.json
```
This looks like the wrong place when running with the integration tests tarball (i.e.: CI pipelines that don't build from source).  I think we want something based on TARGET_DIR here.  The path within the tarball of this file can be under dist/target/ instead of integration_tests so this relative path still works even in the tarball.

jihoonson replied:
A couple of questions for my understanding.

> CI pipelines that don't build from source

Are these different ones from the pre-merge CI that has been running for this PR? Can you point me to where I can find about them? I'd like to know more about those CI pipelines.

> I think we want something based on TARGET_DIR here.

`TARGET_DIR` is set to `"$SCRIPTPATH"/target` in this script file. Is it overwritten to somewhere else somehow in the CI pipelines you mentioned?

> The path within the tarball of this file can be under dist/target/ instead of integration_tests so this relative path still works even in the tarball.

IIUC, you seem to be suggesting the path should be `"${TARGET_DIR}"/../../dist/target/spark-rapids-default-configs.json`? Would this work though when you run the integration test locally from the `integration_tests` directory? This is currently recommended in https://github.com/NVIDIA/spark-rapids/tree/branch-24.10/integration_tests#running.

jihoonson replied:
I talked to @gerashegalov offline and got answers 🙂

jihoonson replied:
I have updated the path as suggested.

jlowe commented on integration_tests/src/assembly/bin.xml:
2024 copyrights

jihoonson commented:
I updated the base branch as this PR is now targeting the next release.

jihoonson commented:
Thanks to @gerashegalov, the integration test module now generates the config file by itself. The config generation logic is self-contained in the integration test module and we no longer have the issue described in https://github.com/NVIDIA/spark-rapids/pull/11283#issuecomment-2269987155. This change has made in https://github.com/NVIDIA/spark-rapids/pull/11283/commits/b04484a98c4040c91a84b7869dd55e4f824ab579.

jihoonson commented on integration_tests/src/assembly/bin.xml:
```diff
@@ -50,6 +50,11 @@
+        <file>
+            <!--This should point to the proper `dist` directory for the scala version being used-->
+            <source>${project.basedir}/../dist/target/spark-rapids-default-configs.json</source>
+            <outputDirectory>integration_tests</outputDirectory>
```
Oops, my apologies. I have fixed this in https://github.com/NVIDIA/spark-rapids/pull/11283/commits/3e59847ead0489a7a0176edbd405f4f9815a1269.

jlowe commented on integration_tests/src/main/scala/com/nvidia/spark/rapids/tests/DumpDefaultConfigs.scala:
```diff
@@ -0,0 +1,35 @@
+ */
+object DumpDefaultConfigs {
+  def main(args: Array[String]): Unit = {
+    if (args.length < 2) {
```
We not only want to fail when they don't provide enough arguments but also when they provide too many, otherwise we're probably not doing what they expect.
```suggestion
    if (args.length != 2) {
```

jlowe commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:
```diff
@@ -2523,6 +2552,50 @@ val SHUFFLE_COMPRESSION_LZ4_CHUNK_SIZE = conf("spark.rapids.shuffle.compression.
+    val PLAIN, JSON = Value
+  }
+
+  def dumpConfigsWithDefault(args: Array[String]): Unit = {
```
Nit: This should take separate arguments, have `main` deal with pulling them out to call this properly rather than hiding the call signature here.

jihoonson replied:
Yeah that would be better, but this logic cannot be in the `DumpDefaultConfigs.main()` as the `withResource` arm currently has a dependency on some spark stuff, which triggers shimming.

jlowe replied:
I'm not asking for the _logic_ to be in main, I'm asking for main to pass the arguments separately.  Main just has to unpack the strings from the args[] array rather than having dumpConfigsWithDefault take an unclear and error-prone argument list.

jlowe reviewed (approved):
lgtm other than method signature nit which is not must-fix.

jihoonson commented:
Build

jihoonson commented:
Can someone help to run the CI? I don't seem to have enough power for that.

gerashegalov commented:
> Can someone help to run the CI? I don't seem to have enough power for that.

Create a PR to add your GH user here https://github.com/NVIDIA/spark-rapids/blob/8ef2e41d4b4126eecf9130043a2b7bedf2f4263f/.github/workflows/blossom-ci.yml#L78

gerashegalov commented:
build

jihoonson commented:
The CI is failing because some tests in `hash_aggregate_test` are assuming `spark.sql.legacy.allowNegativeScaleOfDecimal` to be set to false by default, which is good as this is exactly what we want to catch and fix with this PR 🙂 Though, those tests (`test_hash_grpby_sum` and `test_hash_grpby_sum_full_decimal`) seem to intend to test certain cases with decimals with negative scales, so I think we should keep the negative decimal scale enabled for these tests. In https://github.com/NVIDIA/spark-rapids/pull/11283/commits/c4ba1a241e6382c39ada1c2067e2d26be38ba893, I added `conf = copy_and_update(conf, {'spark.sql.legacy.allowNegativeScaleOfDecimal': 'true'})` to overwrite this property in each test. I'm not quite sure if this is the best way to do so though, please let me know.

gerashegalov commented on integration_tests/pom.xml:
```diff
@@ -78,6 +88,31 @@
+                            <goal>run</goal>
+                        </goals>
+                        <configuration>
+                            <target>
+                                <taskdef resource="net/sf/antcontrib/antcontrib.properties"/>
+                                <java classname="com.nvidia.spark.rapids.tests.DumpDefaultConfigs"
+                                    failonerror="true"
+                                    classpathref="maven.compile.classpath"
+                                    fork="true">
+                                    <arg value="json"/>
+                                    <arg value="${project.build.directory}/spark-rapids-default-configs.json"/>
+                                </java>
+                            </target>
```
Curious why we end up not needing this https://github.com/NVIDIA/spark-rapids/pull/11283#issuecomment-2265948274

jihoonson replied:
Good question. I'm not quite sure why it works. I can add those java options here just in case..

gerashegalov reviewed (approved):
LGTM with one caveat:

Just double-checking: 
beyond propagating the classpath, part of the reason why we looked into forking out a JVM for config dump execution is to propagate extra options for JDK17.   https://github.com/NVIDIA/spark-rapids/pull/11283#issuecomment-2265948274

jihoonson commented:
> It is necessary to set fork=true, if the class being launched by this task or any libraries being used by that class, call APIs like java.lang.System.exit() or java.lang.Runtime.exit().

I have noticed that the `RapidsConf` task in `dist/pom.xml` should set `fork=true` as well per https://ant.apache.org/manual/Tasks/java.html since now it can call `System.exit()`. Fixed this in https://github.com/NVIDIA/spark-rapids/pull/11283/commits/45beca389221c8c5535ba6800ba1e4355dcaa6f1.

jihoonson commented:
> I have noticed that the `RapidsConf` task in `dist/pom.xml` should set `fork=true` as well per https://ant.apache.org/manual/Tasks/java.html since now it can call `System.exit()`. Fixed this in [45beca3](https://github.com/NVIDIA/spark-rapids/commit/45beca389221c8c5535ba6800ba1e4355dcaa6f1).

NVM. Java 8 seems to fail to have some classpath issue with this change. It could be nice to fix, but I don't think the change that adds `System.exit()` in `RapidsConf.main()` is necessary for this PR, so I just reverted it.

jihoonson commented:
build

jihoonson commented:
build

jihoonson commented:
Hi @jlowe, I have fixed all tests and the CI is green now. Could you have another look?

jlowe commented on integration_tests/src/main/python/arithmetic_ops_test.py:
```diff
@@ -876,8 +907,7 @@ def test_acosh(data_gen):
 def test_columnar_acosh_improved(data_gen):
     assert_gpu_and_cpu_are_equal_collect(
-            lambda spark : unary_op_df(spark, data_gen).selectExpr('acosh(a)'),
-            {'spark.rapids.sql.improvedFloatOps.enabled': 'true'})
```
Why are we removing this?  The test is explicitly trying to test when this config is true.  Removing is also inconsistent with other tests below that explicitly set this to true.

jihoonson replied:
I removed it because `spark.rapids.sql.improvedFloatOps.enabled` is now set to true by default. I can add it back if we want to make it explicit.

jihoonson replied:
Well, I agree it's better to keep it. Added this and another back in https://github.com/NVIDIA/spark-rapids/pull/11283/commits/04d86b8f596daf9a0743894cf165ace55af13007.

jlowe commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:
```diff
@@ -2531,12 +2531,6 @@ val SHUFFLE_COMPRESSION_LZ4_CHUNK_SIZE = conf("spark.rapids.shuffle.compression.
-      System.err.println(s"Usage: ${this.getClass.getCanonicalName}" +
-        s" {common_configs_path} {advanced_configs_path}")
-      System.exit(1)
-    }
```
I don't think this should have been completely removed.  I can see avoiding System.exit, but why not throw an exception with a clear error message?  Removing this means we'll get something like ArrayIndexOutOfBoundsException later on which is not as helpful for debugging.

jihoonson replied:
I agree that this is good to have, but don't agree that we should do it in this PR. This is not existing code, but is what I added in this PR as I thought it's nice to add, and removed because of the compile error.

jihoonson replied:
Created https://github.com/NVIDIA/spark-rapids/issues/11333.

jihoonson replied:
We can. I just think that it would be better to do that in a separate PR given that I have already reverted this code.

jlowe replied:
Oh right, it used to be related and then the design changed again.  Sure, I'm fine deferring this.

jihoonson replied:
Yeah, there have been some changes going back and forth. Thank you for the review and guiding me to the right direction!

jihoonson commented:
build
